### ISSUE TITLE
Gradient Accumulation Error with Ulysses Sequence Parallel Causes Inconsistent Loss

### ISSUE THREAD

**Issue by GoneZ5** [verl-project/verl#2857]
First of all, thank you for your great work on this repository!
I encountered a loss inconsistency issue when using Ulysses Sequence Parallel for SFT (commit id: [2c85b432996ad28ed9756ae745e7fcb7ec6eee10](https://github.com/volcengine/verl/commit/2c85b432996ad28ed9756ae745e7fcb7ec6eee10)). In the verl framework, enabling Ulysses SP will automatically normalize the train batch size, so that gradient accumulation is used to keep the global batch size unchanged (I think this is a great design). 

```
def _normalize_config_bsz(self):
    dp_size = self.device_mesh.size(0) if not self.ulysses_device_mesh else self.ulysses_device_mesh.size(0)
    if self.device_mesh.get_rank() == 0:
        print(f"Normalize batch size by dp {dp_size}")

    assert self.config.data.train_batch_size % dp_size == 0, f"Global batch size {self.config.data.train_batch_size} is not divisible by dp size {dp_size}"

    self.config.data.train_batch_size //= dp_size

    assert self.config.data.train_batch_size % self.config.data.micro_batch_size_per_gpu == 0
```

However, when I ran SFT experiments with the same data and SP set to 1, 2, and 8, I observed large differences in loss among the runs. The loss and grad_norm curves are as follows:

- Purple: sft-qwen3-32b-lr5e-5-32k-gpu64-bsz64-sp1
- Green: sft-qwen3-32b-lr5e-5-32k-gpu64-bsz64-sp2
- Blue: sft-qwen3-32b-lr5e-5-32k-gpu64-bsz64-sp8

<img width="256" height="192" alt="Image" src="https://github.com/user-attachments/assets/8176ccf9-d828-459a-b2e0-3f0fea1794ff" />

<img width="256" height="192" alt="Image" src="https://github.com/user-attachments/assets/dc7d1e96-f28d-4e3d-a86e-8d17284a7293" />

Additionally, I modified the batch size normalization logic so that when SP is enabled, the global batch size is directly reduced by a factor of SP, and no gradient accumulation is performed. 
```
self.config.data.train_batch_size //= (dp_size * getattr(self.config, "ulysses_sequence_parallel_size", 1))
```
I then re-ran experiments with SP=1 and SP=2. In order to keep the global batch size consistent, I used twice as many GPUs for SP=2. In this case, the loss curves were consistent across both experiments. The new loss and grad_norm curves are as follows:

- Black: sft-qwen3-32b-lr5e-5-32k-gpu32-bsz32-sp1
- Pink: sft-qwen3-32b-lr5e-5-32k-gpu64-bsz32-sp2

<img width="256" height="192" alt="Image" src="https://github.com/user-attachments/assets/b54738d3-a075-48bc-9051-64e6bcad2040" />

<img width="256" height="192" alt="Image" src="https://github.com/user-attachments/assets/bb81b62e-a88a-4ac2-99e9-9a0c9f624405" />

My questions are:

- Is the loss inconsistency caused by incompatibility between SP > 1 and gradient accumulation, or is it a problem with gradient accumulation itself?

- If the incompatibility is indeed due to SP > 1 and gradient accumulation, will this also affect RL experiments, such as GRPO or DAPO?

I look forward to your reply, and thank you again for your excellent work on this repository!

**Comment by puneeshkhanna** [comment]
https://github.com/volcengine/verl/issues/2919 - Suggested a fix in an issue which I raised. Maybe that would fix your issue

**Comment by GoneZ5** [comment]
> [#2919](https://github.com/volcengine/verl/issues/2919) - Suggested a fix in an issue which I raised. Maybe that would fix your issue

Thanks for your reply! I have tried passing `n_micro_batches` as a parameter to the `_compute_loss_and_backward` method and performing `loss /= n_micro_batches` before `loss.backward()`, but the loss still has a large error. The loss and grad_norm curves are as follows:

- Black: sft-qwen3-32b-lr5e-5-32k-gpu32-bsz32-sp1
- Pink: sft-qwen3-32b-lr5e-5-32k-gpu64-bsz32-sp2
- Green: sft-qwen3-32b-lr5e-5-32k-gpu32-bsz32-sp2-ga2 (original batch size normalization code)

<img width="256" height="192" alt="Image" src="https://github.com/user-attachments/assets/f4c4e1e7-3777-4346-a9ef-d4b3e1b3f442" />

<img width="256" height="192" alt="Image" src="https://github.com/user-attachments/assets/ebb9f5af-3b3b-4b53-80f9-63b53560b933" />

**Comment by puneeshkhanna** [comment]
Hope you removed the division by n micro batches in below line too for step loss logging  
```
for micro_batch in micro_batches:
            loss = self._compute_loss_and_backward(batch=micro_batch, n_micro_batches=n_micro_batches) #/ n_micro_batches
            step_loss += loss.item()
```

**Comment by GoneZ5** [comment]
> Hope you removed the division by n micro batches in below line too for step loss logging
> 
> ```
> for micro_batch in micro_batches:
>             loss = self._compute_loss_and_backward(batch=micro_batch, n_micro_batches=n_micro_batches) #/ n_micro_batches
>             step_loss += loss.item()
> ```

Yes, I removed it.

**Comment by puneeshkhanna** [comment]
Also I think for this line, we should pass grad_scaler to False. But your run showing quite different loss with grad accumulation.
`loss = gather_outpus_and_unpad(loss, gather_dim=0, unpad_dim=0, padding_size=pad_size, grad_scaler=False)`

**Comment by puneeshkhanna** [comment]
Maybe try above too, it has a chance to help.

**Comment by GoneZ5** [comment]
Yup, I noticed this before and tried it, but the loss was still inconsistent. But I didn't try it with performing `loss /= n_micro_batches` before `loss.backward()`. I'll try it again. Thank you!

**Comment by GoneZ5** [comment]
But I think we should keep `grad_scaler=True`, because the loss of experiment **sft-qwen3-32b-lr5e-5-32k-gpu64-bsz32-sp2** is correct.

**Comment by Ber666** [comment]
Hi, thanks for sharing your experiment results! We are also running SFT with sp and gradient accumulation, and the trained model shows some strange behaviors. Have you figured out the root cause of this inconsistent loss?

**Comment by GoneZ5** [comment]
Hi, thanks for the follow-up.
After the ablation runs, I’m afraid I haven’t been able to pinpoint the root cause for the inconsistent loss.

**Comment by Schilings** [comment]
> Hi, thanks for the follow-up.嗨，感谢您的跟进。 After the ablation runs, I’m afraid I haven’t been able to pinpoint the root cause for the inconsistent loss.消融运行后，恐怕我无法查明丢失不一致的根本原因。


Regarding the loss calculation during gradient accumulation: if the loss is computed using a token-level strategy where the number of valid tokens can vary across different micro-batches, wouldn't scaling the loss simply by 1 / num_micro_batches be inaccurate?
<img width="1143" height="434" alt="Image" src="https://github.com/user-attachments/assets/e6091704-83e5-4f76-a271-18138819012c" />

**Comment by exiarepairii** [comment]
> > Hi, thanks for the follow-up.嗨，感谢您的跟进。 After the ablation runs, I’m afraid I haven’t been able to pinpoint the root cause for the inconsistent loss.消融运行后，恐怕我无法查明丢失不一致的根本原因。Hi, thanks for the follow-up.嗨，感谢您的跟进。 After the ablation runs, I’m afraid I haven’t been able to pinpoint the root cause for the inconsistent loss.消融运行后，恐怕我无法查明丢失不一致的根本原因。
> 
> Regarding the loss calculation during gradient accumulation: if the loss is computed using a token-level strategy where the number of valid tokens can vary across different micro-batches, wouldn't scaling the loss simply by 1 / num_micro_batches be inaccurate?关于梯度累积过程中的损失计算：如果损失是采用令牌级策略计算的，且不同微批次的有效令牌数量可能不同，那么仅用1/微批次数量来缩放损失会不会不准确呢？ <img alt="Image" width="1143" height="434" src="https://private-user-images.githubusercontent.com/85475959/487274204-e6091704-83e5-4f76-a271-18138819012c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgwMjY4MjksIm5iZiI6MTc1ODAyNjUyOSwicGF0aCI6Ii84NTQ3NTk1OS80ODcyNzQyMDQtZTYwOTE3MDQtODNlNS00Zjc2LWEyNzEtMTgxMzg4MTkwMTJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTE2VDEyNDIwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2NmUxOTA1MWE2Nzc0ZGZjNzVlNGEyNDBkMmEyYWYzMWNjN2E1NDdhODUzOTE0NWRlYzc2YjE2NGI1OGQ3N2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.UujeoxC0rojVKk7pbgg1iFYSIR6bCQgT4chNHcIzznE">

i think so
not only sft, but also ppo training methods implemented in verl face this problem
see https://huggingface.co/blog/gradient_accumulation

**Comment by Taishi-N324** [comment]
> not only sft, but also ppo training methods implemented in verl face this problem
> see https://huggingface.co/blog/gradient_accumulation

I am testing with DAPO, and it seems that the behavior also changes in DAPO.

<img width="1656" height="916" alt="Image" src="https://github.com/user-attachments/assets/b75396ff-429f-4971-af1b-6aa679baa3da" />

<img width="828" height="458" alt="Image" src="https://github.com/user-attachments/assets/4d4d0045-b06d-45d5-b8d1-9b5b430811ff" />

<img width="1656" height="916" alt="Image" src="https://github.com/user-attachments/assets/61f3d965-f90a-4872-b3a4-f05a5bb68771" />

**Comment by puneeshkhanna** [comment]
With balance dp token and divide by n micro batches fix, the implementation in SFT is same as recommended here https://huggingface.co/docs/accelerate/en/usage_guides/gradient_accumulation#gradient-accumulation-on-training-samples-of-variable-size and https://github.com/huggingface/accelerate/blob/main/examples/by_feature/gradient_accumulation_for_autoregressive_models.py

**Comment by longxudou** [comment]
https://github.com/volcengine/verl/commit/e160d3b2e0e0ed7607b6def2c7cac4ddba6b8d6f PR from @puneeshkhanna fixed this problem.

We might close this issue.

**Comment by Kirrito-k423** [comment]
定时清理检查结论：按社区 issue 定时处理规则关闭此 issue。

证据：最近一次非当前运行账号的公开回复或引用是 2025-11-12T09:16:17Z（actor: puneeshkhanna, kind: mentioned），早于本轮 6 个月阈值 2026-01-13T01:10:17.218155+00:00。
检查来源：https://github.com/verl-project/verl/issues/2857，comments API https://api.github.com/repos/verl-project/verl/issues/2857/comments，timeline API https://api.github.com/repos/verl-project/verl/issues/2857/timeline。

如果作者或社区成员仍能提供新的复现、证据或需求背景，请在这里重新评论请求维护者重开，或创建新的 issue 并链接到本 issue。

Scheduled cleanup conclusion: this issue is being closed under the community issue processing policy.

Evidence: The latest public reply or reference from an account other than the current automation account was 2025-11-12T09:16:17Z (actor: puneeshkhanna, kind: mentioned), which is older than the six-month threshold for this run (2026-01-13T01:10:17.218155+00:00).
Sources checked: https://github.com/verl-project/verl/issues/2857, comments API https://api.github.com/repos/verl-project/verl/issues/2857/comments, timeline API https://api.github.com/repos/verl-project/verl/issues/2857/timeline.

If the author or community can still provide a fresh reproduction, evidence, or requirement context, please comment here to ask maintainers to reopen it, or open a new issue and link back to this one.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
